### PR TITLE
[REG 2.084] hashOf will fail to compile for some structs/unions that recursively contain shared enums

### PR DESCRIPTION
**Pull request by n8sh**

### CODE REVIEW

File: test/hash/src/test_hash.d [hunk 1]
```diff
@@ -232,6 +233,19 @@ void issue20034()
...
+void issue21642() @safe nothrow pure
+{
+    enum C : char { _ = 1, }
+    union U { C c; void[0] _; }
+    shared union V { U u; }
+    cast(void) hashOf(V.init);
+    // Also test the underlying reason the above was failing.
+    import core.internal.convert : toUbyte;
+    shared C c;
+    cast(void) toUbyte(c);
```
Comment (PetarKirov) [inline test/hash/src/test_hash.d:246]: Instead of casting to void here and above, it's better to check if the return value is the expected one.

Comment (n8sh) [reply test/hash/src/test_hash.d]: Added a check here. Above there is no particular value the hash code is supposed to be.

Comment (PetarKirov) [reply test/hash/src/test_hash.d]: My motivation is the question: why would an optimizing compiler not remove a call to a pure function, if it's result is not used?

Technically such optimization happens after semantic analysis (where the bug manifests), in the backend, but in theory the frontend could also discard such function calls, so it's better to make the code more robust.

**Review by RazvanN7 — APPROVED** [review]

File: test/hash/src/test_hash.d (hunk 1 above)
Comment (n8sh) [reply test/hash/src/test_hash.d]: Is there a change you would suggest that wouldn't make the test brittle?

Comment (RazvanN7) [reply test/hash/src/test_hash.d]: @PetarKirov Pure functions that may throw exceptions should not be optimized away. This pattern is prevalent in druntime: a function that does some checks and throws if the input is not the expected one.

**Comment by RazvanN7** [comment]
@n8sh This seems to fail the tester pipeline. Any update on that?

File: src/core/internal/convert.d [hunk 2]
```diff
@@ -769,7 +769,7 @@ const(ubyte)[] toUbyte(T)(const ref T val) if (is(T == enum))
     if (__ctfe)
     {
         static if (is(T V == enum)){}
-        return toUbyte(cast(const V) val);
+        return toUbyte(*cast(const V*) &val);
```
Comment (RazvanN7) [inline src/core/internal/convert.d:772]: This change is reported to not be covered by any tests.

Comment (n8sh) [reply src/core/internal/convert.d:772]: The coverage report doesn't take into account the `tests` directory. It is covered by the new test added in this PR.

**Comment by n8sh** [comment]
Accidentally rebased on master instead of stable. Fixing.

**Pull request merged**